### PR TITLE
Adding visual captchas for register and login.

### LINKED ISSUE #638: Captcha for signups
This was already discussed in #124, but I think it deserves a seperate issue.

One option that was mentioned is [https://www.hcaptcha.com/](https://www.hcaptcha.com/), but that one looks terrible to me, with the main focus on monetisation and no mention of privacy or accesilibity issues.

There is also [mtcaptcha](https://www.mtcaptcha.com/) which looks slightly better as they mention GDPR compliancy and proper accesiblity. But its still not great because we rely on a company.

What I would prefer is an open source option that we can embed directly in Lemmy. I havent looked for that, but there should be some projects that we could use.

### PR DESCRIPTION
**Pull request by dessalines**
Fixes #638

### CODE REVIEW

**Comment by Nutomic** [comment]
Cool! We actually had a discussion about this on Matrix recently, and I think there are some very good reasons to make this a separate project. Most importantly it would allow other projects to use the captcha as well, and hopefully contribute to it. Plus it would help to reduce the amount of code and compilation time.

**Comment by dessalines** [comment]
A rust one that's separate from lemmy already exists, made by the same person who does this library I'm using :) https://github.com/daniel-e/rust-captcha

I thought about using it, but I realized what it does is pretty easy, and its probably better to bake it into lemmy, than have yet another service running on docker.

**Comment by dessalines** [comment]
Okay this is ready for you to test. It works well here, luckily alpine linux has `espeak`, so this was pretty easy.

**Comment by Nutomic** [comment]
- If you fail a captcha, it keeps showing the same one. Should generate a new one instead.
- After pressing the play button for audio, it should be replaced with a pause button. Instead, clicking it again will play the audio twice.
- You should exclude characters like `l`/`I` and `0`/`O`, because they are really hard to tell apart.
- I dont think we need a captcha for login.
- We need a config option to enable/disable the captcha (maybe also an option to enable/disable captchas for login).
- Might be helpful to have a button for "generate new captcha" if the current one is unreadable.

**Comment by dessalines** [comment]
> If you fail a captcha, it keeps showing the same one. Should generate a new one instead.

Good call... I also need to invalidate that uuid on the back end after a failed check.

> After pressing the play button for audio, it should be replaced with a pause button. Instead, clicking it again will play the audio twice.

I can do this.

> You should exclude characters like l/I and 0/O, because they are really hard to tell apart.

What about making it all numbers? I was thinking about this for internationalization. Unfortunately I'd have to generate this now, rather than using its default random character generators. But then I could get rid of the whole capital letter, lowercase letter stuff.

> I dont think we need a captcha for login.

I can remove.

> We need a config option to enable/disable the captcha (maybe also an option to enable/disable captchas for login).

Good call, I can add one.

> Might be helpful to have a button for "generate new captcha" if the current one is unreadable.

This would be easy to do.

**Comment by dessalines** [comment]
Okay @Nutomic I added all these, except for the character changing, because unfortunately the captcha library doesn't support custom characters yet. I opened up an issue over there tho: https://github.com/daniel-e/captcha/issues/6

**Comment by Nutomic** [comment]
- The pause button doesnt do anything
- If you enter the captcha wrong, it generates a new one but the value you entered before is still there, would be nice to clear that (same when you click the "reload captcha" button)
- Its still not clear whether its case sensitive or not
- In fact I cant solve a single captcha, even when I definitely enter the same characters and same case
- Pressing enter in the captcha text field should submit, but loads a new captcha instead
- Captcha length seems completely random, it should be consistent so everyone gets the same difficulty
- Logs are completely filled with the base64 data, to the point that I cant even scroll back to actual log entries anymore. Either you need to exclude that from logs, or serve it as normal files instead.
- Also, why use base64 strings and not just `vec<u8>`?

File: server/src/api/user.rs [hunk 1]
```diff
@@ -439,6 +479,56 @@ impl Perform for Oper<Register> {
...
+    _pool: &DbPool,
+    websocket_info: Option<WebsocketInfo>,
+  ) -> Result<Self::Response, LemmyError> {
+    let captcha_settings = Settings::get().captcha;
+
+    if !captcha_settings.enabled {
+      return Ok(GetCaptchaResponse {
+        png: "disabled".to_string(),
+        uuid: "disabled".to_string(),
+        wav: None,
```
Comment (Nutomic) [inline server/src/api/user.rs:497]: This is very hacky, should return an error message instead.

File: server/src/lib.rs [hunk 2]
```diff
@@ -211,9 +214,56 @@ where
...
+
+  // Building proper speech text for espeak
+  for mut c in captcha.chars() {
+    let new_str = if c.is_alphabetic() {
+      if c.is_lowercase() {
+        c.make_ascii_uppercase();
+        format!("lower case {} ... ", c)
+      } else {
+        c.make_ascii_uppercase();
+        format!("capital {} ... ", c)
```
Comment (Nutomic) [inline server/src/lib.rs:228]: This will need localisation.

Comment (dessalines) [reply server/src/lib.rs:228]: Lets just hold off on that until I can get the captcha maintainers to allow custom text, then we can just use numbers instead.... or just all lower case so we don't have to deal with it.

File: server/src/api/user.rs (hunk 1 above)
Comment (dessalines) [reply server/src/api/user.rs:497]: I'll think of another way to do this. My option are either giving an error field, or just wrapping the whole thing in an option.

**Comment by dessalines** [comment]
> The pause button doesnt do anything

I disabled the button, it really seems more work than is necessary to write a whole functioning audio player in javascript just to play a captcha. I'll see if its that much work to add a pause event listener.

> Its still not clear whether its case sensitive or not

I'm not sure how I would show that, the letters themselves are a mix of upper and lower case.

> In fact I cant solve a single captcha, even when I definitely enter the same characters and same case

I'm not sure what this means, it does check for the correct case.

> Pressing enter in the captcha text field should submit, but loads a new captcha instead

I can see if I can do.

> Captcha length seems completely random, it should be consistent so everyone gets the same difficulty

I can't make this consistent unfortunately, the captcha library only has an `add_char()`, without being able to set the initial number of chars, or say which char you want to add. This should wait until they allow custom setting of chars.

> Logs are completely filled with the base64 data, to the point that I cant even scroll back to actual log entries anymore. Either you need to exclude that from logs, or serve it as normal files instead.

Hrmm I could've sworn I got rid of all logs here.. If its always logged on loglevel verbose, I might have to add an exclusion.

> Also, why use base64 strings and not just vec<u8>

I convert from Vec<u8> to base64, because that's what the front end needs. For some reason doing base64 was trivial in rust, but doing it on the front end was a lot more difficult.

**Comment by dessalines** [comment]
I'll change some of these in a little bit.

File: server/src/lib.rs (hunk 2 above)
Comment (Nutomic) [reply server/src/lib.rs:228]: Okay that also works.

**Comment by Nutomic** [comment]
> I disabled the button, it really seems more work than is necessary to write a whole functioning audio player in javascript just to play a captcha. I'll see if its that much work to add a pause event listener.

Alright, then just show a greyed out play button or something like that.

> I'm not sure what this means, it does check for the correct case.

It seems the check is wrong somehow, because even correct captchas werent accepted. I can look into this if you cant reproduce.

Anyway I think we need to wait for the custom character option in the library before merging this.

**Comment by dessalines** [comment]
Wait, so no captchas worked for you? I tested this and it did... but I'll test again.

**Comment by Nutomic** [comment]
I tested at least 3, and was certain that I got them right (skipped some difficult ones).

**Comment by dessalines** [comment]
Ok these are fixed, except for the capital letters / custom chars bits. I also commented logging the send messages (they were cluttering up the logs anyway, and its better to read those on the front end anyway). And I couldn't figure out a good way to do this with env_logger, because the actual logging is in the message send, which happens on the same line for all messages.

I think we should still merge, even before the custom chars because:

- You can easily disable captchas in the `lemmy.hjson`.
- I have no idea how long it'll take for https://github.com/daniel-e/captcha/issues/6 to get looked at, it might be a while, based on the activity of the other tickets. And when it does, I'll get notified. 

I tested a bunch of captchas, and they all worked. If they are a little difficult, maybe we could change the default difficulty to easy?

**Comment by Nutomic** [comment]
Okay now I can solve the captchas without problem, no idea what was going on before. I also tried with different difficulty settings, but the captchas look pretty much identical to me. The captcha layout is pretty messy, so that could use some cleanup.

But all in all, this is good to merge imo.

**Pull request merged**